test/testscript.go

```diff
@@ -32,27 +32,70 @@ func NewProjectTest(t *testing.T, projectDir string) *ProjectTest {
 	}
 }
 
+func copyFile(srcPath string, dstPath string, mode os.FileMode) error {
+	srcContent, err := os.ReadFile(srcPath)
+	if err != nil {
+		return fmt.Errorf("reading source file %s: %w", srcPath, err)
+	}
+
+	if err := os.WriteFile(dstPath, srcContent, mode); err != nil {
+		return fmt.Errorf("writing destination file %s: %w", dstPath, err)
+	}
+	return nil
+}
+
+func makeDir(dirPath string, mode os.FileMode) error {
+	if err := os.MkdirAll(dirPath, mode); err != nil {
+		return fmt.Errorf("creating directory %s: %w", dirPath, err)
+	}
+	return nil
+}
+
 func (e *ProjectTest) CopyFile(srcRelPath string) *ProjectTest {
 	setupFunc := func(env *testscript.Env) error {
 		src := filepath.Join(e.projectDir, srcRelPath)
 		dst := filepath.Join(env.WorkDir, srcRelPath)
 		dstDir := filepath.Dir(dst)
-		if err := os.MkdirAll(dstDir, 0755); err != nil {
-			msg := "failed to create destination directory %s: %w"
-			return fmt.Errorf(msg, dstDir, err)
+
+		if err := makeDir(dstDir, 0755); err != nil {
+			return err
 		}
+		return copyFile(src, dst, 0644)
+	}
+	e.setupFuncs = append(e.setupFuncs, setupFunc)
+	return e
+}
 
-		srcContent, err := os.ReadFile(src)
-		if err != nil {
-			return fmt.Errorf("failed to read source file %s: %w", src, err)
+func (e *ProjectTest) CopyDir(srcRelPath string) *ProjectTest {
+	setupFunc := func(env *testscript.Env) error {
+		srcDir := filepath.Join(e.projectDir, srcRelPath)
+		dstDir := filepath.Join(env.WorkDir, srcRelPath)
+
+		walkFunc := func(srcPath string, info os.FileInfo, err error) error {
+			if err != nil {
+				return fmt.Errorf("walking '%s': %w", srcPath, err)
+			}
+
+			relPath, err := filepath.Rel(srcDir, srcPath)
+			if err != nil {
+				return fmt.Errorf("creating relative path: %w", err)
+			}
+			dstPath := filepath.Join(dstDir, relPath)
+
+			if info.IsDir() {
+				return makeDir(dstPath, info.Mode())
+			} else {
+				return copyFile(srcPath, dstPath, info.Mode())
+			}
```

Comment:
Do you want me to hardcode the permissions like `CopyFile` or is it ok to use the permissions of the existing directories and files? What if we want to copy an executable file?

Reply:
Seems like keeping the original permissions might be the move. I was not sure if there is any reason to avoid it (tests will be run in other environments/runners, with secrets being written to files), but at the moment, looking at this again, it seems fine?

Reply:
updated to preserve file and directory permissions when copying